### DISCUSSION TITLE
Project Update

### DISCUSSION THREAD

**Discussion by jeremystretch** [netbox-community/netbox#5853] [General]
I wanted to share a quick update regarding the status of the NetBox project. Since its open source release in 2016, I have had the privilege of serving as its lead maintainer as part of my full-time employment, initially by DigitalOcean and then briefly with Network to Code. Unfortunately, since my departure from NTC, the amount of time I’m able to spend maintaining the project has decreased drastically.

NetBox is currently down to two maintainers, @dansheps and myself (plus @ryanmerolle, our community manager). We are challenged with handling around a hundred new issues every month on a project with thousands of active users. Whereas in the past we’ve worked hard to ensure every issue is afforded consideration, this is no longer a realistic possibility. Going forward, please understand that your bug reports and especially feature requests may go unanswered. Likewise, future NetBox releases will become far less frequent and focus on addressing critical bugs rather than the introduction of new functionality.

For organizations requiring guaranteed support for NetBox, I have started offering paid professional support on a 12-month term through my company, [NetVerity](https://netverity.dev/). (Unfortunately we are not able to offer support services on an hourly basis.) It is my hope that securing enough customers will allow me to continue supporting NetBox long-term, however this is still a very new initiative.

**Comment by xkilian** [comment]
Thank you for the update and the investment you have done in the project. There are many medium and larger enterprises that use and depend on the quality product that Netbox has become. I can say as a member of a company that is benefitting from Netbox, we are at the point where we were looking at commercial support for Netbox.
Like a lot of companies we have internal ressources, we could be self supporting, but I find that it is important to support companies that invest in supporting, pro services  AND developping their products. I hope others that have benefited and will continue to benefit will support more explicitely the entities behind Netbox.

**Comment by dejantep** [comment]
Sad news about this amazing project. Haven't seen this one coming.

**Reply by jeremystretch** [reply]
Unfortunately it's a fact of life to deal with in the open source world. Previously my work on NetBox was sponsored by Network to Code, but they have since opted to discontinue their financial support of the project.

It's not all doom and gloom though. We're looking at potential new sponsors to help maintain a consistent frequency of work on the project, and I've started work on new features for v2.11 already. It likely just won't come as early as it would have otherwise.

**Reply by raschkeb** [reply]
Is this related to their (NtC) fork of Netbox, Nautobot?

I just started to promote Netbox as IPAM/DCIM with my employer, now I am unsure what to focus on, to be honest.

**Comment by pseudonode** [comment]
@jeremystretch Any thoughts on a donation page? I am happy to contribute as I have benefited from NetBox's existence, and I also think others will too.

**Reply by jeremystretch** [reply]
The other maintainers and I have discussed it, but we've identified a few reasons sponsorship isn't very tenable for this project:

1. NetBox is provides value primarily to the companies who use it, not their employees. I wouldn't feel right accepting money from individuals who don't benefit from the project directly.
2. Most companies don't have a way to budget or account for "small," discretionary donations. If you need to pay $20K for a license to use a product, no problem. But tossing a grand or two into an open source project when there's no obligation is a hard sell.
3. People often conflate donations with the right to demand certain features or changes. It's best to simply avoid the potential conflict if we can.
4. Some of the other maintainers, for various reasons, are unable to accept financial compensation for their work.
5. _Any_ time money is involved, you need to have rules on how to handle it, and ensure that those rules are followed consistently. It adds a significant amount of overhead to the project.

**Comment by cpmills1975** [comment]
@jeremystretch Firstly, thank you for the incredible effort you have put in to NetBox over the years, it's an awesome tool.

The Nautobot fork troubles me. Reading through the release notes for v1.0.0 I fail to believe it hasn't been in the pipeline for some time. It seems like they have taken many of the feature requests from NetBox, implemented them and rather than contributing them back to NetBox, have saved them up for a rebadged release of their own. Clearly there are/were some politics at play here that we mere users aren't aware of.

I thought I saw yesterday that NetBox had been forked under your NetVerity brand - but now can't find it so I'm guessing it has been made private? With the offer of commercial support and bespoke development (I can't blame Jeremy, a man has to eat), and the announcement above of reduced frequency of NetBox bug/feature approvals, is the apache licensed NetBox effectively dead in the water at this point?

As a champion of this tool, do I need to pitch the purchase of a support contract from NetVerity or do I need to consider jumping ship to NTC if spending money isn't an option?

What about outstanding feature requests/bug reports? If you search for my name in the issue list, you will see I have a couple raised and have offered to fix a few (which I'll continue with). Are my feature requests now unlikely to ever be accepted? Should I be raising similar requests to Nautobot to see if they gain any traction there - ultimately, I just want a tool that works for me and my colleagues, I don't really care who implements it!

Last point. Shame about the name. Nautobot hardly rolls off the tongue does it!

**Reply by jeremystretch** [reply]
> I thought I saw yesterday that NetBox had been forked under your NetVerity brand - but now can't find it so I'm guessing it has been made private?

Nah, that's just me messing around with GitHub. (There are ~1400 forks of NetBox on GitHub currently.)

> is the apache licensed NetBox effectively dead in the water at this point?

Not at all. We released v2.10.5 yesterday, and I'm working on v2.11 as we speak.

> As a champion of this tool, do I need to pitch the purchase of a support contract from NetVerity or do I need to consider jumping ship to NTC if spending money isn't an option?

NetBox isn't going anywhere. I can't share any details at the moment but I'm exploring several options in addition to the paid support model.

> What about outstanding feature requests/bug reports? If you search for my name in the issue list, you will see I have a couple raised and have offered to fix a few (which I'll continue with). Are my feature requests now unlikely to ever be accepted?

Some feature requests get picked up right away, some languish for a bit, and some die off. (We a few have feature requests that are several _years_ old.) You can review our [issue intake policy](https://github.com/netbox-community/netbox/wiki/Issue-Intake-Policy) on the wiki. Once an issue has been tagged as `needs review`, it's fair game. If you've volunteered for a tagged issue and not yet been assigned, please ping me in the comments.

**Reply by cpmills1975** [reply]
Thank you for your swift reply @jeremystretch, it's reassuring to know that the tool I use and evangelise about so much isn't going anywhere any time soon.

Time will tell of course whether the right tool for me and my users is NetBox or Nautobot and I'm yet to have that conversation with my user base. To cover all bases, I've raised similar FR's as my open ones here against Nautobot. At this point in the development of both NetBox and Nautobot, it would be trivial for me to raise similar PRs against both. Clearly as the code bases diverge, this will be less easy and we will of course have to choose which path to follow moving forward.

**Comment by xkilian** [comment]
> The Nautobot fork troubles me. Reading through the release notes for v1.0.0 I fail to believe it hasn't been in the pipeline for some time. It seems like they have taken many of the feature requests from NetBox, implemented them and rather than contributing them back to NetBox, have saved them up for a rebadged release of their own. Clearly there are/were some politics at play here that we mere users aren't aware of.

This is what irks me, the politics, and the fact that ultimately the users will lose out on future collaboration and current resolved backlog. There are many fine additions to Nautobot that were planned originaly in NetBox and should be backported in Netbox, as well as ideas that should be implemented. I agree that this was planned a while ago.
- Consolidation of export templates/scripts into jobs (planned in 2.11)
- Hiding UI elements based on permissions (suggested in UI rework ideas)
- Ability to launch or register jobs in plugins and exposing this in the API (good idea)
- More statuses with sane defaults for all items. This was such a heated debate in Netbox, and in the end users just want a little more flexibility. This is one of the things has driven us crazy in our current system, lack of states/statuses!
- Data validation plugin API; Making sure data quality is maintained and is a key component that is not currently well addressed. Great idea in Nautobot.
- Git as a data source, is a good idea for those doing automation and for maintaining jobs/scripts/reports
- Relationship model which seems to address one of my blockers (modeling L2 SPB or VXLANs) though the idea to permit this for other types of relastionships in addition to vlans is interesting. (planned in 2.11 in some form, but not as a new model)
- custom fields on all models (Was this in the pipeline? I think so..)

All great ideas, just makes me sad that it has come to this. I hope some NetBox developpers/contributors/new devs will take a look at some of the features and implement them back in NetBox. The more similar the code, the more there can be benefits for all. If features are accepted and companies dedicate some development time to implement a feature or two it will benefit all. 
I also thank @jeremystretch for all his work and continued support of the projet. I do hope more developpers and companies will put some money or dev time to further the project and that more feature requests are accepted and flagged for help-wanted. We all benefit from it! Making our jobs and our collegues jobs a little less tedious or difficult.
If I had one wish, is that NetBox was just a wee bit more open to use-cases that help users without denaturing the product or impacting long term goals and maintainability. This will help onboard new developpers and contributors.
I say all this the most respectfully possible. (though in written form it can be a challenge)
Cheers,

**Reply by jeremystretch** [reply]
I think this is a really fair, well-constructed take. Thank you for sharing!

> If I had one wish, is that NetBox was just a wee bit more open to use-cases that help users without denaturing the product or impacting long term goals and maintainability.

There's a fundamental mantra in open source: "No is temporary, yes is forever," meaning that you can always kick a feature request down the road, but once you've committed to something you're more or less stuck with it. I've really taken this to heart, admittedly, at times, to a fault. I am more often than not the most conservative voice in any discussion regarding the introduction of new features, though I swear my intentions are good. I'd rather delay the execution of an idea until the time is right than risk a subpar implementation, or worse, a blocking change. I'm particularly vulnerable to these decisions because ultimately it's my responsibility to make things work.

I'll give an example. For a long time, one of the most requested features in NetBox was the ability to define custom interface types. My take was and is that there are a finite number of discrete physical interface types in the world, and it makes no sense for people to create their own unless they're physically manufacturing bespoke hardware. I caught a lot of flak for this until we debuted the [device type library](https://github.com/netbox-community/devicetype-library), a curated collection of predefined device types and their related components. Suddenly, the method behind the madness became apparent: Because interface types are all standardized in NetBox, anyone can write a YAML file defining `type: 1000base-t` and it will _just work_ for everyone else.

Custom fields are another example. For a long time, support for custom fields on additional objects (particularly interfaces) has been a top request. It's difficult to convey to people not intimately familiar with the code base why the entire implementation needed to be overhauled before we expanded it any further: It works for some things already, why not add more? But in v2.10 I was finally able to knock that out, and now custom fields are both better performing and [much easier to work with](https://netbox.readthedocs.io/en/stable/release-notes/version-2.10/#reimplementation-of-custom-fields-4878). And now we're slated to expand custom fields to (almost) all other models in v2.11. (In fact, it's largely already been done.)

None of this is to suggest that I know exactly how things will unfold in the future, or even that I have a fully-formed plan at any given time, but simply to share my perspective. I admit that I often don't do a great job with this, but it's often a struggle just to keep up with the sheer volume of discussion.

Regarding the fork, I can only say that my priorities for the project and those of folks at NTC clearly did not align. Their focus, in my experience, is heavily skewed toward supporting customization and plugins, which isn't a bad thing. But it needs to be balanced with ensuring the long-term sustainability of the project; work which is often neither particularly fun nor rewarding. There's also a ton of work that remains around simply extending the data model: wireless networks, virtual circuits, overlays, etc.

In such cases where priorities don't agree, my advice has always been to maintain a local fork of NetBox with your own modifications. When executed with care, this allows for the best of both worlds: You can continue to benefit from NetBox updates, and one day we might be able to merge your changes into core. NTC opted to go a different way, creating a permanently diverging fork and redistributing it under their own brand. It's disappointing, but at this point not worth dwelling on.

**Reply by cpmills1975** [reply]
Thanks for this @jeremystretch. It feels like it came from the heart and your intentions sound a lot like the approach Apple takes to it's product development - it's better to let others be first to market and get it wrong and wait until you can do it well. I've certainly noticed that some of your questions around feature requests, probing how the user "thinks" something will work show a desire to get a deep understanding of how the feature would work in the real world before jumping head-long in to implementing something sub-par. Would it be worth you putting together a blog post somewhere to outline where you think NetBox should go over the next 3-12 months?

Having played with the sandboxed demo of Nautobot 1.0.0b (and immediately causing it to throw an exception simply by leaving the name field on a device blank 🙄), I do worry that it isn't quite ready for prime time yet - perhaps too much focus has been put on getting too much out of the door on day one without adequately testing any of it. I have to say, for all the features I'd like to see in NetBox, those that are there are nicely stable!

Ultimately I guess users will vote with their feet and while a network automation tool sounds on paper like a good idea, I think my preference would still be for a solid source of truth at the core which provides good hooks out to things like Ansible Tower etc where the automation can be run on tooling better suited to it rather than one product to rule them all.

**Reply by DanSheps** [reply]
> I think my preference would still be for a solid source of truth at the core which provides good hooks out to things like Ansible Tower etc where the automation can be run on tooling better suited to it rather than one product to rule them all.

Just speaking about my own personal goals/views, I think honestly,  this is the best way to do it in the end.  If you wanted to run everything in NetBox, you could in theory point a Webhook back to a script and launch a script that does your automation.  This would keep it somewhat decoupled from NetBox.

**Comment by baldgeek** [comment]
https://blog.networktocode.com/post/why-did-network-to-code-fork-netbox/

I read that as there was a philosophical difference in where netbox 
developers wanted to go and where NtC wanted to go with netbox. I see 
both sides, and I am curious to see where things are in 6-12 months.

**Reply by xkilian** [reply]
I read through the roadmap and the "why" post. I can understand some divergence in where to go that would lead to the change. NtC has done amazing quality stuff over the years too! I have to say, that I am more along the lines of Source of Truth+automation around documentation+data integration and less about intent based configuration/automation. Each company will have their use-cases.

Though some of the ideas put forward make sense for both, IMO.

- UI improvemets 
- circuit modeling(supporting path through ptp circuits), 
- custom fields, 
- more web hooks, 
- custom statuses, 
- organization models, 
- the plugins, validation, etc. 

In any case there is a massive market for this type of product and I wish both entities much success that can be shared back. (at a minimum as ideas, as code, etc.). Though I will still say I am irked that they "held on" to improvements that could have benefited both code bases and that were not material to the "differentiation". Obviously understanding that as the code bases diverged this would be more difficult.

**Reply by cpmills1975** [reply]
It seems much of NTC's additional functionality could have been released as a series of plugins with supporting work to the core where required to facilitate a tighter integration in to the UI or additional hooks in to the existing templates etc. NTC could have continued to sell their automation/integration services off the back of those plugins and benefitted from a maintained core.

**Reply by dgarros** [reply]
I wish this was true but it isn't. 
Forking has always been the last option on the table and we carefully evaluated all other options including developing the features that our customers are looking for as plugins. Since 2.8 has been released we developed dozens of plugins some public, some private and I think at this point we have a really good understanding of what you can do with a plugin and what you can't do with a plugin in NetBox. We discussed how to improved the plugin API with the NetBox maintainers multiple times but again we had a different vision on that the plugin API should look like. Ultimately some of the features that are very important to us must be implemented in the core.

**Comment by tb-killa** [comment]
Hi @jeremystretch thank you very much for your info.
Too bad to read all this here.
Maybe you can still answer the following info ?

You don't work for NeTworktoCode anymore ?
NTC will not invest any more work in Netbox itself ?
Does this also mean @lampwins is no longer a maintainer for Netbox itself but concentrates only on "nautobot" ?

Do you plan to partially merge "nautobot" commits if they are profitable ?
